libs/mailroom/common/src/lib/interceptors/log-to-database/log-to-database.interceptor.ts

```diff
@@ -4,35 +4,46 @@ import { InjectRepository } from '@nestjs/typeorm';
 import { tap } from 'rxjs/operators';
 import { Repository } from 'typeorm';
 
-import { MailroomEmail } from '../../entities/all.entities';
+import { MailroomAttachment, MailroomEmail } from '../../entities/all.entities';
+import { EmailStatus, IMailroomEmailOutbound, ITamuRelayResponse } from '../../types/mail.types';
 
 @Injectable()
 export class LogToDatabaseInterceptor implements NestInterceptor {
   @InjectRepository(MailroomEmail)
   public repo: Repository<MailroomEmail>;
 
+  @InjectRepository(MailroomAttachment)
+  public attachmentsRepo: Repository<MailroomAttachment>;
+
   public intercept(context: ExecutionContext, next: CallHandler) {
-    const request = context.switchToHttp().getRequest();
-    console.log('LogToDatabaseInterceptor', request.body);
+    return next.handle().pipe(
+      tap(async (value: { response: ITamuRelayResponse; email: IMailroomEmailOutbound }) => {
+        // create a save the email / relay response to db
+        const outbound = await this.repo
+          .create({
+            relayResponse: JSON.stringify(value.response),
+            text: value.email.text,
+            subject: value.email.subject,
+            to: value.email.to,
+            from: value.email.from,
+            deliveryStatus: value.response.accepted.length > 0 ? EmailStatus.Accepted : EmailStatus.Rejected
+          })
+          .save();
 
-    // Hasn't been through the pipe to transform request into MailroomOutbound
-    // const email: Partial<MailroomEmail> = {
-    //   content: 'Joe mamma',
-    //   from: 'joemamma@gmail.com',
-    //   deliveryStatus: EmailStatus.Accepted
-    // };
-    // this.repo.create(email).save();
+        // if attachments are present, save each one to the db
+        if (value.email.attachments) {
+          const attachments: Array<Express.Multer.File> = value.email.attachments;
 
-    return next.handle().pipe(
-      tap((response) => {
-        // response here contains a MailroomOutbout object (or should)
-        // const email: Partial<MailroomEmail> = {
-        //   content: 'Joe mamma',
-        //   from: 'joemamma@gmail.com',
-        //   deliveryStatus: EmailStatus.Accepted
-        // };
-        // this.repo.create(email).save();
-        console.log('LogToDatabaseInterceptor inside of the handle observable', response);
+          attachments.forEach((attachment) => {
+            this.attachmentsRepo
+              .create({
+                email: outbound,
+                mimeType: attachment.mimetype,
+                blob: attachment.buffer
+              })
+              .save();
+          });
+        }
```

Comment:
This is async and will not necessarily run in the same event loop. Put in array and promisify them.

Reply:
Since nothing depends on these being saved to the db, is it all that important?

Reply:
If that's the case, no. Either way, something to keep an eye out for because in other places where it does matter it can cause race conditions.